### PR TITLE
iEEG: Different ways to start `SEEG/ECOG Implantation`

### PR DESCRIPTION
**Pull request by chinmaychinara91**
This PR refactors the **SEEG/ECOG Implantation** option:

1. Users can **right click on a CT or MRI** individually and start implantation. Only show MRI Viewer and load iEEG panel.

2. Users can now **right click on the Subject** both in **Anatomy** and **Functional** modes, and use the GUI above to decide on what modalities they want to proceed with. 
Screenshot 2024-11-19 125245
-  `MRI`: Start implantation on the default MRI. Only show MRI Viewer and load iEEG panel. This is similar **to right click on a MRI and start implantation**.
- `CT`: Start implantation on the CT. If multiple CTs then UI pops up for users to choose. Only show MRI Viewer and load iEEG panel. This is similar **to right click on a CT and start implantation**.
- `MRI+CT`: Start implantation with CT overlayed on default MRI. If multiple CTs then UI pops up for users to choose. Only show MRI Viewer and load iEEG panel.
- `MRI+CT+Isosurface`: Start implantation with CT overlayed on default MRI. Since there can be only one isosurface per subject as of now, based on which CT file the isosurface was generated from, it loads that CT. If the associated CT is not found then UI pops up for users to choose. Show MRI Viewer, show 3D fig and load iEEG panel.

Special case (@rcassani correct me here if I am wrong):
- `CT+Isosurface`: This is a case when users would just import raw CTs and start their implantation without having an MRI. Results will be same as `MRI+CT+Isosurface` case just that MRI viewer will just have CT and 3D fig will have the isosurface+CT3D.

Relevant forum topics:
- https://neuroimage.usc.edu/forums/t/electrode-location-using-brianstrom-of-verison-25-oct-2024/48747/10

### CODE REVIEW

**Review by rcassani — CHANGES_REQUESTED** [review]
While the PR works, the organization of the code can be greatly improved.

There are GUI interaction in different parts, including plotting, which should not be the case

EDIT: This is the diff for all the changes that I added:
https://github.com/brainstorm-tools/brainstorm3/compare/547a300..17a68e2

File: toolbox/gui/panel_ieeg.m [hunk 1]
```diff
@@ -3042,9 +3042,12 @@ function CenterMriOnElectrode(sElec, hFigTarget)
 
 
 %% ===== CREATE IMPLANTATION =====
-function CreateImplantation(MriFile) %#ok<DEFNU>
+% 
+function CreateImplantation(MriFile, isAsk) %#ok<DEFNU>
+    %
```
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/gui/panel_ieeg.m]: Solved in abfa1caa051ca87ea64d52a93f7c83252c343343

File: toolbox/anatomy/tess_isosurface.m [hunk 2]
```diff
@@ -165,7 +165,7 @@
     
     % Save isosurface
     sMesh.Comment = sprintf('isoSurface (ISO_%d)', isoValue);
-    sMesh = bst_history('add', sMesh, 'threshold_ct', 'CT thresholded isosurface generated with Brainstorm');
+    sMesh = bst_history('add', sMesh, 'threshold_ct', ['Thresholded CT: ' sMri.FileName]);
```
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/anatomy/tess_isosurface.m]: Added in 1df44554574f1d45890cfd42b140a7ca1abf2277

File: toolbox/gui/panel_ieeg.m [hunk 3]
```diff
@@ -3083,6 +3086,16 @@ function CreateImplantation(MriFile) %#ok<DEFNU>
...
+
+    % Ask which volume to proceeed with
+    VolMat = struct([]);
+    if isAsk
+        VolMat = AskImplantationVolume(sSubject);
+        if ~VolMat.isMri && ~VolMat.isCt && ~VolMat.isIsosurface
+            return
+        end
+    end
+
```
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/gui/panel_ieeg.m]: Changed in f1b456ad34d778930db6032ceab69eab1ec46a35

File: toolbox/gui/panel_ieeg.m [hunk 4]
```diff
@@ -3152,67 +3223,82 @@ function LoadElectrodes(hFig, ChannelFile, Modality) %#ok<DEFNU>
 end
 
 %% ===== DISPLAY CHANNELS (MRI VIEWER) =====
-% USAGE:  [hFig, iDS, iFig] = DisplayChannelsMri(ChannelFile, Modality, iAnatomy, isEdit=0)
+% USAGE:  [hFig, iDS, iFig] = DisplayChannelsMri(ChannelFile, Modality, iAnatomy, isEdit=0, VolMat=struct([]))
```
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/gui/panel_ieeg.m]: Changed in 44eb30fc24dce9587f2b7e2c1a16391ce72eec70

File: toolbox/gui/panel_ieeg.m [hunk 5]
```diff
@@ -3222,11 +3308,25 @@ function LoadElectrodes(hFig, ChannelFile, Modality) %#ok<DEFNU>
...
+    % (VolMat.isCt && ~VolMat.isMri): user right clicks on subject, chooses 'CT' and it has CT but no MRI
+    % Also say user creates isoSurface from a CT whose dimensions do not match to MRI, just show the CT file for implantation 
+    elseif ~isempty(CtFile) && ((VolMat.isCt && ~VolMat.isMri) || ~isequal(size(sMri.Cube(:,:,:,1)), size(bst_memory('LoadMri', CtFile).Cube(:,:,:,1))))
+        [hFig, iDS, iFig] = view_mri(CtFile, [], [], 2);
+    % Case-1: User right clicks on subject, chooses 'CT+MRI' or 'CT+MRI+Isosurface' then always overlay CT on MRI
+    % Case-2: User right clicks on subject and it has MRI but no CT ('CtFile' = []). Only MRI opens.
+    % Case-3: User right clicks on subject, with MRI present, if user deleted the CT(s) ('CtFile' = []) but not the Isosurface. Only MRI opens.
+    else
+        [hFig, iDS, iFig] = view_mri(MriFile, CtFile, [], 2);
+    end
```
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/gui/panel_ieeg.m]: Changed in 44eb30fc24dce9587f2b7e2c1a16391ce72eec70

```
% == DISPLAY THE MRI VIEWER
    if length(MriFiles) == 1
        [hFig, iDS, iFig] = view_mri(MriFiles{1}, [], [], 2);
    else
        [hFig, iDS, iFig] = view_mri(MriFiles{1}, MriFiles{2}, [], 2);
    end
```

File: toolbox/gui/panel_ieeg.m [hunk 6]
```diff
@@ -3152,67 +3223,82 @@ function LoadElectrodes(hFig, ChannelFile, Modality) %#ok<DEFNU>
...
+        if VolMat.isIsosurface
+            ProtocolInfo = bst_get('ProtocolInfo');
+            iIsosurface = find(cellfun(@(x) ~isempty(regexp(x, '_isosurface', 'match')), {sSubject.Surface.FileName}));
+            sSurf = load(bst_fullfile(ProtocolInfo.SUBJECTS, sSubject.Surface(iIsosurface).FileName));
+            if isfield(sSurf, 'History') && ~isempty(sSurf.History)
+                % Search for CT threshold in history
+                ctEntry = regexp([sSurf.History{:, 3}], 'Thresholded CT:\s*(\S+)', 'tokens', 'once');
+                if ~isempty(ctEntry)
+                    % Locate corresponding CT volume in Anatomy
+                    iCtFile = find(cellfun(@(x) ~isempty(regexp(x, ctEntry{1}, 'match')), {sSubject.Anatomy.FileName}));
```
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/gui/panel_ieeg.m]: Keep it simple
[~, iCtIso] = ismember(ctEntry{1}, {sSubject.Anatomy.FileName});

Changed: 44eb30fc24dce9587f2b7e2c1a16391ce72eec70

File: toolbox/gui/panel_ieeg.m [hunk 7]
```diff
@@ -3155,7 +3155,7 @@ function CreateImplantation(MriFile) %#ok<DEFNU>
                 % Return intersection of the found and then update iCtVol
                 if ~isempty(ctEntry)
                     [~, iCtIso] = ismember(ctEntry{1}, {sSubject.Anatomy.FileName});
-                    if ~isempty(iCtIso)
```
Comment (chinmaychinara91) [inline toolbox/gui/panel_ieeg.m]: **Issue-1:**
Screenshot 2024-11-20 151323
Steps:
1. Have multiple CTs and no MRI for a subject.
2. Generate isoSurface from one of the CTs and delete the source CT.
3. Start implantation by right clicking on subject. Select `CT+IsoSurf`. Error comes.

**Issue-2:**
Steps:
1. Have MRI and multiple CTs for a subject
2. Generate isoSurface from one of the CTs and delete the source CT.
2. Generate isoSurface from one of the CTs and delete the source CT.
3. Start implantation by right clicking on subject. Select `MRI+CT_IsoSurf`.
**Result:** The MRI viewer opens up with just MRI and 3D figure with isoSurface+3DMRI.

Expected: In both cases the dialog box for **The CT that was used to create the IsoSurface cannot be found.** should pop up.

Fixed in

**Comment by chinmaychinara91** [comment]
It looks good now. I ran the tests and it seems good to be merged :).

**Pull request merged**